[BUGFIX beta] Ensures that `@each` works with non-arrays and array-likes

chancancode opened the pull request:
`@each` should no-op with non arrays, since there is nothing to track,
and should work with array-likes (anything that implements the
`objectAt` and `length` properties.)

chancancode committed:
[BUGFIX beta] Ensures that `@each` works with non-arrays and array-likes

`@each` should no-op with non arrays, since there is nothing to track,
and should work with array-likes (anything that implements the
`objectAt` and `length` properties.)

TODO: do we intend to support array-like objects or just `isEmberArray`?

Co-authored-by: Chris Garrett <me@pzuraq.com>

chancancode commented:
@pzuraq the assertion further down about `typeof === 'object'` _may_ fall into the same category (but only if it works correctly on release)

pzuraq commented:
That assertion was a direct requirement of the `EachProxy`: https://github.com/emberjs/ember.js/blob/master/packages/@ember/-internals/metal/lib/each_proxy.ts#L128

chancancode commented:
@pzuraq Pushed a second test re: the other assertion. It seems to work on release but double check me. I think there are extra guards somewhere else that prevents it from going into the `EachProxy` code path?

chancancode commented:
Opened https://github.com/emberjs/ember.js/pull/18400 to confirm the tests work.

chancancode commented:
Nevermind! My test was invalid!

wycats commented:
lgtm

chancancode committed:
[BUGFIX beta] Don't suggest to blindly refactor

If the app worked fine given the current semantics, they can probably
just remove the extraneous segments instead of doing a bunch of work
to refactor.

Also removed some deadcode that was not reachable before.

chancancode commented on packages/@ember/-internals/metal/lib/chain-tags.ts:
```diff
@@ -70,48 +70,71 @@ export function getChainTagsForKey(obj: any, path: string) {
+        break;
+      }
 
-      if (segment.indexOf('.') !== -1) {
```
I believe this was dead code.

There are two cases on L93: `segmentEnd === -1` (`foo.@each.bar`) and `segmentEnd !== -1` (`foo.@each.bar.baz`, deprecated).

For the first case, `segment` is `path.slice(lastSegmentEnd, pathLength)` which would be `bar`. It is impossible for this to contain a `.` because we already checked previously.

For the second case, `segment` is `path.slice(lastSegmentEnd, segmentEnd)` which would also be `bar`. It is impossible for this to contain a `.`, because `segmentEnd` is the next occurrence of a `.` after the `@each.`, returned from `indexOf('.', lastSegmentEnd)`.

So in both cases, this conditional cannot ever by true so the branch is effectively dead but we are still doing part of the work (`segment.indexOf('.')`) for no reason.
